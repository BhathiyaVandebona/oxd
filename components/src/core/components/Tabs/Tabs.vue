<template>
  <ul class="tab-list">
    <li class="tab-items" v-for="(tab, i) in tabs" :key="i">
      <span
        tabindex="0"
        class="tab-item"
        @click="setCurrentTabId(tab)"
        :class="{ active: currentTabId == tab.id }"
        @keyup.enter="setCurrentTabId(tab)"
        :id="tab.id"
      >
        <oxd-icon class="tab-icon" :name="tab.icon" v-if="tab.icon" />{{
          $vt(tab.title)
        }}</span
      >
      <div
        class="selected-tab-indicator"
        :class="{ active: currentTabId == tab.id }"
      ></div>
    </li>
  </ul>
  <div class="tab-content">
    <KeepAlive>
      <slot :name="currentTabId"></slot>
    </KeepAlive>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref } from "vue";
import { Tab } from "./types";
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';
import translateMixin from '../../../mixins/translate';

export default defineComponent({
  name: "oxd-tabs",
  props: {
    tabs: {
      type: Array as () => Array<Tab>,
      required: true
    },
  },
  components: {
    "oxd-icon": Icon,
  },
  mixins: [translateMixin],
  setup: function (props, context) {
    const currentTabId = ref(props.tabs[0].id);
    context.emit('selectTabItem', props.tabs[0]);
    return {
      currentTabId
    };
  },
  emits: ['selectTabItem'], 
  methods: {
    setCurrentTabId(tab: Tab) {
      this.currentTabId = tab.id;
      this.$emit("selectTabItem", tab);
    }
  },
});
</script>

<style src="./tabs.scss" lang="scss" scoped></style>
