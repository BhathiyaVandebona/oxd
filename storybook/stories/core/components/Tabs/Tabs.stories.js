import Tabs from '@orangehrm/oxd/core/components/Tabs/Tabs.vue';

export default {
  title: 'Layout/Tabs',
  component: Tabs,
  argTypes: {
    tabs: {
      control: {type: 'Array'},
      table: {
        type: {summary: 'there are three attributes, "id", "title" and the "icon". Icon is not mandatory. i.e [{ id: "tab1", title: "Details", icon : "oxd-posts" }, { id: "tab2", title: "Resume", icon : "oxd-profile" }]'}
      }
    }
  }
};

const Template = args => ({
  setup() {
    return {args};
  },
  components: {'oxd-tabs': Tabs},
  template: '<oxd-tabs v-bind="args"> <template #tab1> <div> Hi this is the content of first tab </div> </template> <template #tab2> <div> Hi this is the content of second tab </div> </template> <template #tab3> <div> Hi this is the content of third tab </div> </template> </oxd-tabs>'
});


const tabsDefault = [
    { id: "tab1", title: "Posts" }, 
    { id: "tab2", title: "Personal Details" }, 
    { id: "tab3", title: "Schedule"}
  ];

const tabsWithIcon = [
    { id: "tab1", title: "Posts", icon : 'oxd-posts' }, 
    { id: "tab2", title: "Personal Details", icon : 'oxd-profile' }, 
    { id: "tab3", title: "Schedule", icon : 'oxd-time'}
  ];

export const Default = Template.bind({});
Default.args = {
        tabs : tabsDefault
};

Default.parameters = {
    docs: {
      source: {
        code:
          '<oxd-tabs \n :tabs=' +
          JSON.stringify(tabsDefault) +
          '\n/>',
      },
    },
  };

export const WithIcon = Template.bind({});
WithIcon.args = {
    tabs : tabsWithIcon
};

WithIcon.parameters = {
    docs: {
      source: {
        code:
          '<oxd-tabs \n :tabs=' +
          JSON.stringify(tabsDefault) +
          '\n/>',
      },
    },
  };
