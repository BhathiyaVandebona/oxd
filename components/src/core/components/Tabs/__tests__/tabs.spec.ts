import {mount, shallowMount} from '@vue/test-utils';
import Tabs from '@orangehrm/oxd/core/components/Tabs/Tabs.vue';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';

describe('Tabs.vue', () => {
    it('Renders two tabs along with icons', async () => {
        const wrapper = mount(Tabs, {
            props: {
                tabs : [ { id: "tab1", title: "Details", icon : 'oxd-posts' }, { id: "tab2", title: "Resume", icon : 'oxd-profile' }]
            },
            slots: {
                tab1: '<div> This is the content of tab 1</div>',
                tab2: '<div> This is the content of tab 2</div>'
            }
        });
        await wrapper.vm.$nextTick();

        expect(wrapper.html()).toMatchSnapshot();
        const icon = wrapper.findComponent(Icon);
        expect(icon.find('.tab-icon')).toBeTruthy();
    });

    it('Renders two tabs without icons', () => {
        const wrapper = shallowMount(Tabs, {
            props: {
                tabs : [ { id: "tab1", title: "Details" }, { id: "tab2", title: "Resume"}]
            },
            slots: {
                tab1: '<div> This is the content of tab 1</div>',
                tab2: '<div>This is the content of tab 2</div>'
            }
        });
        expect(wrapper.find('.tab-icon').exists()).toBeFalsy();
    });

    it('In initial rendering first tab Id is set to the currentTabId', () => {
        const wrapper = shallowMount(Tabs, {
            props: {
                tabs : [ { id: "tab1", title: "Details", icon : 'oxd-posts' }, { id: "tab2", title: "Resume", icon : 'oxd-profile' }]
            }
        });
        expect(wrapper.vm.currentTabId).toBe("tab1");
    });

    it('In initial rendering first tab is set to the selectTabItem by default', () => {
        const wrapper = shallowMount(Tabs, {
            props: {
                tabs : [ { id: "tab1", title: "Details", icon : 'oxd-posts' }, { id: "tab2", title: "Resume", icon : 'oxd-profile' }]
            }
        });
    
        expect(wrapper.emitted('selectTabItem')).toBeTruthy();
        expect(wrapper.emitted('selectTabItem')).toContainEqual(
            [{ 
                id: "tab1", 
                title: "Details", 
                icon : 'oxd-posts' 
            }]
        );});
   
        it('Click on the second tab', () => {
            const wrapper = shallowMount(Tabs, {
                props: {
                    tabs : [ { id: "tab1", title: "Details" }, { id: "tab2", title: "Resume"}]
                },
                slots: {
                    tab1: '<div> This is the content of tab 1</div>',
                    tab2: '<div>This is the content of tab 2</div>'
                }
            });

    
            wrapper.find('#tab2').trigger('click');
            wrapper.vm.$nextTick();
            expect(wrapper.emitted('selectTabItem')).toContainEqual(
                [{ 
                    id: "tab2", 
                    title: "Resume"
                }]
            );
        });

        it('Hit Enter when focusing on the 2nd tab on the first tab', () => {
            const wrapper = shallowMount(Tabs, {
                props: {
                    tabs : [ { id: "tab1", title: "Details" }, { id: "tab2", title: "Resume"}]
                }
            });
            
            wrapper.find('#tab2').trigger('keyup.enter');
            wrapper.vm.$nextTick();
            expect(wrapper.emitted('selectTabItem')).toContainEqual(
                [{ 
                    id: "tab2", 
                    title: "Resume"
                }]
            );
        });
});