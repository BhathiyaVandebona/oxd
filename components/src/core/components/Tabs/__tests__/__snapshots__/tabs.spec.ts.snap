// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs.vue Renders two tabs along with icons 1`] = `
<ul class="tab-list">
  <li class="tab-items"><span tabindex="0" class="tab-item active" id="tab1"><oxd-icon-stub name="oxd-posts" size="medium" class="tab-icon"></oxd-icon-stub>Details</span>
    <div class="selected-tab-indicator active"></div>
  </li>
  <li class="tab-items"><span tabindex="0" class="tab-item" id="tab2"><oxd-icon-stub name="oxd-profile" size="medium" class="tab-icon"></oxd-icon-stub>Resume</span>
    <div class="selected-tab-indicator"></div>
  </li>
</ul>
<div class="tab-content">
  <keep-alive-stub></keep-alive-stub>
</div>
`;
